File: py/spin_glass_nn/hessian_spectrum.py
"""Hessian spectrum analysis."""

from __future__ import annotations
import numpy as np


def eigenvalue_spectrum(diag_H: np.ndarray) -> np.ndarray:
    """Sorted diagonal Hessian curvatures Hkk (not true eigenvalues)."""
    return np.sort(diag_H)


def ipr(eigenvalues: np.ndarray) -> float:
    """Inverse Participation Ratio: IPR = sum(|Hkk| / sum|Hjj|)^2.

    1/N = uniform curvature (glassy); → 1 = few sharp axes (crystallised/ordered).
    """
    vals = np.abs(eigenvalues)
    s = vals.sum()
    if s < 1e-12:
        return 1.0
    return float(((vals / s) ** 2).sum())


def participation_ratio(eigenvalues: np.ndarray) -> float:
    i = ipr(eigenvalues)
    return float(len(eigenvalues)) if i < 1e-12 else 1.0 / i

File: jl/spin_glass_nn/src/HessianSpectrum.jl
module HessianSpectrum

using LinearAlgebra, Statistics

export eigenvalue_spectrum, ipr, participation_ratio

"""
    eigenvalue_spectrum(diag_H)

Given the diagonal Hessian approximation (parameter-axis curvatures Hₖₖ),
return them sorted. These are not true eigenvalues — for the full spectrum
use LinearAlgebra.eigen on the full Hessian.
"""
function eigenvalue_spectrum(diag_H::AbstractVector)
    sort(diag_H)
end

"""
    ipr(eigenvalues)

Inverse Participation Ratio: IPR = Σ(|Hₖₖ| / Σ|Hⱼⱼ|)^2
Range: 1/N (all diagonal entries equal weight, delocalised) to 1 (single dominant axis).
High IPR means the curvature budget is concentrated on few parameter axes:
low IPR ~ broad/uniform curvature (glassy); high IPR ~ few sharp axes (crystallised/ordered).
"""
function ipr(eigenvalues::AbstractVector)
    vals = abs.(eigenvalues)
    s = sum(vals)
    s ≈ 0.0 && return 1.0
    sum((v/s)^2 for v in vals)
end

"""
    participation_ratio(eigenvalues)

PR = 1/IPR — number of "effectively participating" eigenvalues.
"""
function participation_ratio(eigenvalues::AbstractVector)
    i = ipr(eigenvalues)
    i ≈ 0.0 && return Float64(length(eigenvalues))
    1.0 / i
end

end
